README
This is a test repository for debian/ubuntu packages of compiz-0.8.8 modified to work with the MATE desktop environment

NOTE: This is not an official part of MATE.


debian/changelog
compizfusion (0.8.8-4) UNRELEASED; urgency=low

  * Fix slow startup in MATE by reading DESKTOP_AUTOSTART_ID to get the
    smclient id.

 -- Brent Hull <bhull2010@live.com>  Thu, 09 May 2013 9:18:34 -0400

compizfusion (0.8.8-3) unstable; urgency=low

  * Package rename to avoid update prompts for compiz 0.9
  * Add breaks for compiz packages and libdecoration to avoid
    file conflicts with compiz 0.9
  * Now builds without mateconf dependencies
  * Package changed to 3.0 (native) format

 -- Brent Hull <bhull2010@live.com>  Sun, 21 Apr 2013 16:09:14 -0400

compiz (0.8.8-2) unstable; urgency=low

  * Don't remove compiz.desktop, and don't bundle compiz-mate-gtk.desktop
    in debian/mate_files
  * Don't remove compiz-wm.desktop. Install it from package compiz-mate in
    usr/share/mate/wm-properties/compiz-wm.desktop, for marco windows 
    settings to work in compiz.
  * update 001-mate-compiz.patch to update both desktop files to:
    - use X-MATE-WMSettingsModule=marco and X-MateWMSettingsLibrary=marco
    - drop deprecated Encoding=UTF-8 attribute
  * delete no longer needed patch 018_use_marco_settings.patch
  * delete no longer needed patch 060_move_checks_to_compiz.patch, we're
    already checking capabilities in /usr/bin/compiz-mate-gtk wrapper,
    and if we can't do indirect rendering, we fall back to marco.

 -- Jasmine Hassan <jasmine.aura@gmail.com>  Sun, 14 Oct 2012 03:21:07 +0200

compiz (0.8.8-1) unstable; urgency=low

  * new upstream release, non-maintainer fork
  * [debian/compat] Bump to Debhelper v7
  * [debian/copyright] convert to http://dep.debian.net/deps/dep5/ format
  * [debian/control]
    - rename compiz-gtk package to compiz-mate. update description thereof
    - add compiz-locales package, to reduce size of arch-dep compiz-core
    - bump Standards-Version to 3.9.3, Build-Dep debhelper to (>= 7.0.50~)
    - bump Build-Deps compiz-dev and libcompizconfig-dev to current version
    - B-D on dh-autoreconf, drop autoconf automake libtool autotools-dev
    - compiz-core depend on compiz for docs, and update compiz description
    - Remove compiz-kde package (commented out). Disabled via debian/rules
    - hardening-wrapper, to enable it from debian/rules (see below)
    - change the maintainer's email address
  * [debian/source/format] convert to '3.0 (quilt)' source format
  * [debian/patches/*]
  - Add 001-mate-compiz.patch: migration for MATE compatibility.
  - Add 002-mate-compiz_pot_files.patch - pot files MATE changes.
  - All migrated patches assume no files were renamed (ex: by migrate.sh).
    See notes under [debian/rules] further down this changelog.
  - Sync patches from Ubuntu (0.8.6-0u12) that are still needed in 0.8.8
    This is not the full list, only those that were migrated/updated.
    For the rest of (unchanged) patches, see debian/patches/series
    - 013-add-cursor-theme-support.patch: update: migrate to MATE
    - 018_use_marco_settings.patch: rename, migrate, update
    - 021_hide_tooltip_on_decorator.patch: update, migrate to MATE
    - 029_default_options.patch: update w/ new changes over Debian, migrate
      . use new compiz-mate-gtk wrapper script instead of compiz-decorator
      . revert to debian's logo instead of ubuntu's for adherence.
    - 060_move_checks_to_compiz.patch update, refresh
    - 562027-fix-mateconf-ftbfs.patch update, migrate, rename
  - Add 200-linuxmint-logo.patch to replace Debian logo with LinuxMint
  - Sync & update some from Wolfgang Ulrich's Redhat fc16 compiz fork:
    - 102-compiz_windows-decorator.patch: cleaned up, added dep3 header
    - 103-compiz-0.8.6-unloadpluginfix.patch: added dep3 header
  - Add 101-mate-compiz_fix_gwd_window_buttons_placement.patch. Thanks to
    Loftyhauser: http://forums.mate-desktop.org/viewtopic.php?f=8&t=818
  * [debian/patches/series] Indicate all origins of respective patches.
  * [debian/mate_files/*]
    - From Wolfgang Ulrich's forked redhat compiz package:
      - 50-marco-(navigation|system|windows).xml: enables some more compiz
        keybindings based on the marco ones
      - compiz-mate-gtk: wrapper script to run the appropriate decorator
        updated to add --replace options.
    - compiz-mate-gtk.desktop: X-MATE-WMSettingsModule=marco
    - compiz-keybindings.sed: Add keybindings (adapted from 0.8.4-4ubuntu1)
  * [debian/libdecoration0.symbols] after `debuild binary`, generate with:
    dpkg-gensymbols -plibdecoration0 -Pdebian/libdecoration0 | patch -p0
    perl -pi -e 's/-\d.*//' new_symbol_file  #strips the debian version
  * [debian/rules] use simplified dh style, only use necessary overrides
    - override_dh_autoreconf to rename files (gnome->MATE) after patching,
      and override_dh_(clean|builddeb) to revert filenames before dh_clean 
      and dpkg-source --after-build, respectively. The rationale is that we
      can keep a debian/patches much smaller sized, and much more portable!
    - Disable all kde stuff (kconfig/kde/kde4). Remove its xml and schema
    - Import keybindings from marco (adapted from 0.8.4-4ubuntu1)
    - Install compiz-mate-gtk*, remove compiz.desktop
    - Remove libmate(conf|compat) from plugins pkg. They're included in mate
    - Remove core.xml from plugins pkg. It's included in the core pkg
    - override_dh_makeshlibs to fix dependencies on libdecoration0
    - DEB_BUILD_HARDENING=1, to suppress lintian warnings
  * [debian/*.install]
    - compiz-dev: install schemas.xslt file (moved from compiz-core)
    - compiz-core: only install en_GB,en_US,fr locales
    - compiz-locales: all other locales
    - compiz-mate: install gtk files, our new keybindings & wrapper script
    - compiz-plugins: install xml files (moved from compiz-core)
  * [debian/*.docs]
    - Only compiz-core & libdecoration0 have docs. other pkgs use *.links
  * [debian/compiz-mate.links]
    - gtk-window-decorator bin/man to compiz-decorator and compiz-mate-gtk
  * [debian/watch] update download link from freedesktop.org to compiz.org

src/main.c
/*
 * Copyright © 2005 Novell, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software
 * and its documentation for any purpose is hereby granted without
 * fee, provided that the above copyright notice appear in all copies
 * and that both that copyright notice and this permission notice
 * appear in supporting documentation, and that the name of
 * Novell, Inc. not be used in advertising or publicity pertaining to
 * distribution of the software without specific, written prior permission.
 * Novell, Inc. makes no representations about the suitability of this
 * software for any purpose. It is provided "as is" without express or
 * implied warranty.
 *
 * NOVELL, INC. DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
 * NO EVENT SHALL NOVELL, INC. BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
 * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Author: David Reveman <davidr@novell.com>
 */

#ifdef HAVE_CONFIG_H
#  include "../config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>

#include <compiz-core.h>

char *programName;
char **programArgv;
int  programArgc;

char **initialPlugins = NULL;
int  nInitialPlugins = 0;

char *backgroundImage = NULL;

REGION   emptyRegion;
REGION   infiniteRegion;
GLushort defaultColor[4] = { 0xffff, 0xffff, 0xffff, 0xffff };
Window   currentRoot = 0;

int  defaultRefreshRate = 50;
char *defaultTextureFilter = "Good";

Bool shutDown = FALSE;
Bool restartSignal = FALSE;
Bool coreInitialized = FALSE;

CompWindow *lastFoundWindow = 0;
CompWindow *lastDamagedWindow = 0;

Bool replaceCurrentWm = FALSE;
Bool indirectRendering = FALSE;
Bool strictBinding = TRUE;
Bool noDetection = FALSE;
Bool useDesktopHints = FALSE;
Bool onlyCurrentScreen = FALSE;
Bool noFBO = FALSE;
static Bool debugOutput = FALSE;

#ifdef USE_COW
Bool useCow = TRUE;
#endif

CompMetadata coreMetadata;

static void
usage (void)
{
    printf ("Usage: %s\n       "
	    "[--display DISPLAY] "
	    "[--bg-image PNG] "
	    "[--refresh-rate RATE]\n       "
	    "[--fast-filter] "
	    "[--indirect-rendering] "
	    "[--no-detection]\n       "
	    "[--keep-desktop-hints] "
	    "[--loose-binding] "
	    "[--replace]\n       "
	    "[--no-fbo] "
	    "[--sm-disable] "
	    "[--sm-client-id ID] "
	    "[--only-current-screen]\n      "

#ifdef USE_COW
	    " [--use-root-window] "
#endif

	    "[--debug] "
	    "[--version] "
	    "[--help] "
	    "[PLUGIN]...\n",
	    programName);
}

void
compLogMessage (const char   *componentName,
		CompLogLevel level,
		const char   *format,
		...)
{
    va_list args;
    char    message[2048];

    va_start (args, format);

    vsnprintf (message, 2048, format, args);

    if (coreInitialized)
	(*core.logMessage) (componentName, level, message);
    else
	logMessage (componentName, level, message);

    va_end (args);
}

void
logMessage (const char	 *componentName,
	    CompLogLevel level,
	    const char	 *message)
{
    if (!debugOutput && level >= CompLogLevelDebug)
	return;

    fprintf (stderr, "%s (%s) - %s: %s\n",
	      programName, componentName,
	      logLevelToString (level), message);
}

const char *
logLevelToString (CompLogLevel level)
{
    switch (level) {
    case CompLogLevelFatal:
	return "Fatal";
    case CompLogLevelError:
	return "Error";
    case CompLogLevelWarn:
	return "Warn";
    case CompLogLevelInfo:
	return "Info";
    case CompLogLevelDebug:
	return "Debug";
    default:
	break;
    }

    return "Unknown";
}

static void
signalHandler (int sig)
{
    int status;

    switch (sig) {
    case SIGCHLD:
	waitpid (-1, &status, WNOHANG | WUNTRACED);
	break;
    case SIGHUP:
	restartSignal = TRUE;
	break;
    case SIGINT:
    case SIGTERM:
	shutDown = TRUE;
    default:
	break;
    }
}

typedef struct _CompIOCtx {
    int	 offset;
    char *pluginData;
    char *textureFilterData;
    char *refreshRateData;
} CompIOCtx;

static int
readCoreXmlCallback (void *context,
		     char *buffer,
		     int  length)
{
    CompIOCtx *ctx = (CompIOCtx *) context;
    int	      offset = ctx->offset;
    int	      i, j;

    i = compReadXmlChunk ("<compiz><core><display>", &offset, buffer, length);

    for (j = 0; j < COMP_DISPLAY_OPTION_NUM; j++)
    {
	CompMetadataOptionInfo info = coreDisplayOptionInfo[j];

	switch (j) {
	case COMP_DISPLAY_OPTION_ACTIVE_PLUGINS:
	    if (ctx->pluginData)
		info.data = ctx->pluginData;
	    break;
	case COMP_DISPLAY_OPTION_TEXTURE_FILTER:
	    if (ctx->textureFilterData)
		info.data = ctx->textureFilterData;
	default:
	    break;
	}

	i += compReadXmlChunkFromMetadataOptionInfo (&info,
						     &offset,
						     buffer + i,
						     length - i);
    }

    i += compReadXmlChunk ("</display><screen>", &offset,
			   buffer + i, length - 1);

    for (j = 0; j < COMP_SCREEN_OPTION_NUM; j++)
    {
	CompMetadataOptionInfo info = coreScreenOptionInfo[j];

	switch (j) {
	case COMP_SCREEN_OPTION_REFRESH_RATE:
	    if (ctx->refreshRateData)
		info.data = ctx->refreshRateData;
	default:
	    break;
	}

	i += compReadXmlChunkFromMetadataOptionInfo (&info,
						     &offset,
						     buffer + i,
						     length - i);
    }

    i += compReadXmlChunk ("</screen></core></compiz>", &offset, buffer + i,
			   length - i);

    if (!offset && length > i)
	buffer[i++] = '\0';

    ctx->offset += i;

    return i;
}

int
main (int argc, char **argv)
{
    CompIOCtx ctx;
    char      *displayName = 0;
    char      *plugin[256];
    int	      i, nPlugin = 0;
    Bool      disableSm = FALSE;
    char      *clientId = NULL;
    char      *refreshRateArg = NULL;

    programName = argv[0];
    programArgc = argc;
    programArgv = argv;

    signal (SIGHUP, signalHandler);
    signal (SIGCHLD, signalHandler);
    signal (SIGINT, signalHandler);
    signal (SIGTERM, signalHandler);

    emptyRegion.rects = &emptyRegion.extents;
    emptyRegion.numRects = 0;
    emptyRegion.extents.x1 = 0;
    emptyRegion.extents.y1 = 0;
    emptyRegion.extents.x2 = 0;
    emptyRegion.extents.y2 = 0;
    emptyRegion.size = 0;

    infiniteRegion.rects = &infiniteRegion.extents;
    infiniteRegion.numRects = 1;
    infiniteRegion.extents.x1 = MINSHORT;
    infiniteRegion.extents.y1 = MINSHORT;
    infiniteRegion.extents.x2 = MAXSHORT;
    infiniteRegion.extents.y2 = MAXSHORT;

    memset (&ctx, 0, sizeof (ctx));

    for (i = 1; i < argc; i++)
    {
	if (!strcmp (argv[i], "--help"))
	{
	    usage ();
	    return 0;
	}
	else if (!strcmp (argv[i], "--version"))
	{
	    printf (PACKAGE_STRING "\n");
	    return 0;
	}
	else if (!strcmp (argv[i], "--debug"))
	{
	    debugOutput = TRUE;
	}
	else if (!strcmp (argv[i], "--display"))
	{
	    if (i + 1 < argc)
		displayName = argv[++i];
	}
	else if (!strcmp (argv[i], "--refresh-rate"))
	{
	    if (i + 1 < argc)
	    {
		refreshRateArg = programArgv[++i];
		defaultRefreshRate = atoi (refreshRateArg);
		defaultRefreshRate = RESTRICT_VALUE (defaultRefreshRate,
						     1, 1000);
	    }
	}
	else if (!strcmp (argv[i], "--fast-filter"))
	{
	    ctx.textureFilterData = "<default>Fast</default>";
	    defaultTextureFilter = "Fast";
	}
	else if (!strcmp (argv[i], "--indirect-rendering"))
	{
	    /* force Mesa libGL into indirect rendering mode, because
	       glXQueryExtensionsString is context-independant */
	    setenv ("LIBGL_ALWAYS_INDIRECT", "1", True);
	    indirectRendering = TRUE;
	}
	else if (!strcmp (argv[i], "--loose-binding"))
	{
	    strictBinding = FALSE;
	}
	else if (!strcmp (argv[i], "--ignore-desktop-hints"))
	{
	    /* keep command line parameter for backward compatibility */
	    useDesktopHints = FALSE;
	}
	else if (!strcmp (argv[i], "--keep-desktop-hints"))
	{
	    useDesktopHints = TRUE;
	}
	else if (!strcmp (argv[i], "--only-current-screen"))
	{
	    onlyCurrentScreen = TRUE;
	}
	else if (!strcmp (argv[i], "--no-fbo"))
	{
		noFBO = TRUE;
	}

#ifdef USE_COW
	else if (!strcmp (argv[i], "--use-root-window"))
	{
	    useCow = FALSE;
	}
#endif

	else if (!strcmp (argv[i], "--replace"))
	{
	    replaceCurrentWm = TRUE;
	}
	else if (!strcmp (argv[i], "--sm-disable"))
	{
	    disableSm = TRUE;
	}
	else if (!strcmp (argv[i], "--sm-client-id"))
	{
	    if (i + 1 < argc)
		clientId = argv[++i];
	}
	else if (!strcmp (argv[i], "--no-detection"))
	{
	    noDetection = TRUE;
	}
	else if (!strcmp (argv[i], "--bg-image"))
	{
	    if (i + 1 < argc)
		backgroundImage = argv[++i];
	}
	else if (*argv[i] == '-')
	{
	    compLogMessage ("core", CompLogLevelWarn,
			    "Unknown option '%s'\n", argv[i]);
	}
	else
	{
	    if (nPlugin < 256)
		plugin[nPlugin++] = argv[i];
	}
    }

    if (!clientId)
    {
	clientId = getenv ("DESKTOP_AUTOSTART_ID");
    }

    if (refreshRateArg)
    {
	ctx.refreshRateData = malloc (strlen (refreshRateArg) + 256);
	if (ctx.refreshRateData)
	    sprintf (ctx.refreshRateData,
		     "<min>1</min><default>%s</default>",
		     refreshRateArg);
    }

    if (nPlugin)
    {
	int size = 256;

	for (i = 0; i < nPlugin; i++)
	    size += strlen (plugin[i]) + 16;

	ctx.pluginData = malloc (size);
	if (ctx.pluginData)
	{
	    char *ptr = ctx.pluginData;

	    ptr += sprintf (ptr, "<type>string</type><default>");

	    for (i = 0; i < nPlugin; i++)
		ptr += sprintf (ptr, "<value>%s</value>", plugin[i]);

	    ptr += sprintf (ptr, "</default>");
	}

	initialPlugins = malloc (nPlugin * sizeof (char *));
	if (initialPlugins)
	{
	    memcpy (initialPlugins, plugin, nPlugin * sizeof (char *));
	    nInitialPlugins = nPlugin;
	}
	else
	{
	    nInitialPlugins = 0;
	}
    }

    xmlInitParser ();

    LIBXML_TEST_VERSION;

    if (!compInitMetadata (&coreMetadata))
    {
	compLogMessage ("core", CompLogLevelFatal,
			"Couldn't initialize core metadata");
	return 1;
    }

    if (!compAddMetadataFromIO (&coreMetadata,
				readCoreXmlCallback, NULL,
				&ctx))
	return 1;

    if (ctx.refreshRateData)
	free (ctx.refreshRateData);

    if (ctx.pluginData)
	free (ctx.pluginData);

    compAddMetadataFromFile (&coreMetadata, "core");

    if (!initCore ())
	return 1;

    coreInitialized = TRUE;

    if (!disableSm)
	initSession (clientId);

    if (!addDisplay (displayName))
	return 1;

    eventLoop ();

    if (!disableSm)
	closeSession ();

    coreInitialized = FALSE;

    finiCore ();
    compFiniMetadata (&coreMetadata);

    xmlCleanupParser ();

    if (initialPlugins)
        free (initialPlugins);

    if (restartSignal)
    {
	execvp (programName, programArgv);
	return 1;
    }

    return 0;
}
